Create diagnostic apps for VlasovSpecies and GkSpecies

manauref opened the pull request:
Now the diagnostics are managed by a diagnostic App (see App/Diagnostics/SpeciesDiagnostics). This will allow us to remove >90% of the diagnostic-related code currently in species, reuse more code, and make implementing other diagnostics easier (I think).

Major changes I'd like to keep:

1. Now in input files, instead of having something like
```
   diagnosticMoments = {"M0", "M2"},
   diagnosticIntegratedMoments = {"intM0","intM2"},
```
we simply specify all the species diagnostics in a flat table as
```
   diagnostics = {"M0", "M2","intM0","intM2"},
```
2. The names of the diagnostics are no longer tied to the names implemented in DistFuncMomentCalc. For that reason, and given that the above change will require people to update their input files anyway, I've taken the opportunity to do the following renaming:
- Vlasov: ```u``` -> ```Udrift``` and ```vtSq```->```VtSq```.
- GK: remove all "Gk" prefixes, so diagnostics now are consistent with Vlasov: M0, M1, M2, VtSq, Upar, etc.
- Also in GK: rename ```GkEnergy``` -> ```HE``` for Hamiltonian energy since we were outputting ```intHE``` which is the volume integral of the hamiltonian energy.
In general I kept the diagnostic moments capitalized, so it plays nicely with its volume integrated diagnostic, e.g. Beta & intBeta.

3. Source diagnostic are now specified through a `diagnostics={...}` table in the Source app. At present Gyrofluid and Gyrokinetic sources recycle the species diagnostics, so one can request the same diagnostics of the source as we do of the species (e.g. intM0 would mean the source's intM0). These produce files named like `electron_source_intM0.bp`.

4. I've introduced a number of functions to handle deltaF calculations, primarily in GK, so that we don't have if-statements checking if we are doing a deltaF sim all over the place. When doing fullF those functions are simply empty methods. In the process of doing this some orders of operations changed. This causes some if not all the deltaF, fuctuationBCs or perturbed diagnostic regression tests to fail. But I checked them by eye and they looked ok.

5. Also in the process of pruning support for deltaF calculations I renamed `f0` -> `fBackground`. And output files that previously had in its name  `f0` now have `background`, or if they had `f1` now they have `fluctuation`.

6. In GK we previously had an option to compute perturbed diagnosticMoments, but the diagnosticIntegratedMoments remained full-F. Now all the diagnostics will be perturbed if `"perturbed"` is included in `diagnostics={...}`. If one wishes the full-f integrated diagnostics one can add the contribution due to the constant background in post-processing.

Changes I'd like feedback on/would be willing to alter:

7. Some diagnostics depend on other diagnostics. For example, M2Thermal depends on M2 and M2Flow. Previously there was an inconsistency in that sometimes dependencies caused some diagnostics to write out the things they depend on, but not aways. So as of now this branch writes out all the things that diagnostic depends on. I repeat: it writes out all the intermediate diagnostics computed. That means that if you request M2Thermal you'll also write out M2 and M2Flow. For this reason a large number of regression tests fail (I've checked almost every single regression test, sometimes by eye), because new files are now created.

8. intHE in GK GenGeo regression tests will fail because I'm computing it slightly differently than before. I checked that if I compute it exactly the same as before it doesn't fail. @nmandell look at the commented code for this diagnostic in GkDiagnostics. The difference is between dividing or multiplying by the Jacobian.

Two additional notes:

9. At the moment the integrated moments of the sources are accumulated/integrated in time. Although not done here I propose that we don't do this. Greg mentioned to @tnbernard in a call at some point that it is easier to visualize the volume integrated terms in the PDE (as done in the FluidvGK paper comparing GDB and Gkeyll for Helimak), not the volume and time integrated terms.

10. Due to the new files mentioned in 7, and some changes in order of operations (primarily for deltaF GK calculations which now use some new methods to avoid endless checking of the same if-statements), nearly every regression test will fail. I've checked every one of them, I believe, so I'm confident we can accept the new results.

Lastly, boundary fluxes are not included in this merge. I realized that this was a big enough change for now. BC apps and boundary flux diagnostics I'll do in a separate PR. For that reason (and to not interfere with neutral diagnostics) most of the old infrastructure for diagnostics is still there, but I'll eventually delete it.

tnbernard reviewed (commented):
#1-2 seem reasonable. #3 is a good change in order to have consistency of diagnostics between Vlasov and Gk sources. Previously intSrcM0 was automatically written for Gk but not Vlasov. As I don't use perturbed diagnostics I defer to other reviewers' assessments on those changes. Future diagnostics pertaining to #9 would indeed be useful.

JunoRavin commented:
First of all, I want to publicly document how awesome I think all of this is. I think this is a really impressive re-factoring @manauref 

Some thoughts based on number orderings:
1-5. This all seems fine to me (though I defer to @nmandell on any thoughts he has on certain aesthetic choices; I am fine with the Vlasov aesthetic choices and like the refactoring of source diagnostics).
6. I liked before that the key-value pair in the table was perturbed = true. Admittedly I never run with this on, but this made sense to me from a design perspective. Could you comment/expand upon your decision to change this to what looks like a string evaluation as opposed to the old perturbed=true? By default are the background diagnostics written out as well so that one can do a simple ev command to add the two quantities together?
7. I think this is fine; however, given recent experience by all of us (@tnbernard included) dealing with our particular I/O strategy stressing Lustre file systems, we should be careful. Obviously all of us are way over-subscribed, but if we are now entering territory where a user may write out more diagnostics than they expected it is worth *strongly* considering the revamping of diagnostics to write to a single "grid diagnostics" and "integrated diagnostics" set of files. I defer to you @manauref, @nmandell, and @pcagas as I believe (from my conversations with y'all) we are now capable of writing multiple data structures to the same .bp file, encoding within the .bp file what is what, and allowing postgkyl to process the multi-diagnostic files. However, actually doing this work will probably have to wait.

No comments on 8 or 10

9. Can you elaborate on 9 somewhere? Either here or over Slack or the Gkeyll listserv? I don't fully follow what the proposal is. All the integrated diagnostics for Vlasov are just volume integrated and then written out at each point in time (to my knowledge). Are y'all doing more sophisticated things on the GK side?

pcagas commented:
I confirm that postgkyl fully supports specifying variable names for a bp file.

manauref commented:
Responses to @JunoRavin :

6. Under the hood simply passing `"perturbed"` instead of `perturbed=true` is easier, since the latter mixes index-value pairs and key-value pairs. We could do `perturbed=true` but I didn't feel terribly compelled. Most simulations are full-F so this isn't the number one use-mode.
When using `"perturbed"` all diagnostics are delta-F and no full-F or background-F diagnostics are computed. We write out the background, so the user can compute diagnostics of the background in post-processing. But we currently do not calculate the background-F diagnostics. We could consider implementing that though (I'm happy to do that when/if someone needs it).
7. I gave some thought to how to write out multiple diagnostics to a single file. Yes, the simplest thing would be to do what you suggested: all grid diagnostics in one file and all integrated diagnostics in another file. I can add that if you guys want, it wouldn't be hard (I think it'll involve passing an extra flag to the diagnostics table, something like `"groupDiagnostics"`.
9. I think @nmandell at some point chose to make source integrated diagnostics integrated in time also (so new volume integrated values are not just appended to the DynVector, they are accumulated). I'm just proposing we don't do this for volume int<...> diagnostics.

JunoRavin commented:
With regards to the group-ed diagnostics, is there any reason we should not make this the default behavior just so you don't have to include logic on whether to group the diagnostics? Obviously it requires updating documentation (certainly the case for Vlasov..) and it has the potential to be a bit of a system shock change.

But by and large this is a significant improvement if postgkyl can handle this change. It reduces file system stress, makes it easier for us to navigate the directory where we run simulations, and there is no performance penalty (in fact the I/O performance may improve). 

The only reason we wouldn't make this the default behavior is if we're concerned not having separate diagnostic files will impact our work-flows as we struggle to adapt to new pgkyl commands to fetch our preferred moment from within the grid or integrated diagnostics file.

manauref commented:
We could make it the default. We could have a vote/survey in Slack so
everyone gets a change to voice their opinion. I would say the other
important factor is whether the systems like more smaller files or fewer
big files, and you and others probably have more knowledge about this.

On Thu, 27 May 2021 at 11:32, James Juno ***@***.***> wrote:

> With regards to the group-ed diagnostics, is there any reason we should
> not make this the default behavior just so you don't have to include logic
> on whether to group the diagnostics? Obviously it requires updating
> documentation (certainly the case for Vlasov..) and it has the potential to
> be a bit of a system shock change.
>
> But by and large this is a significant improvement if postgkyl can handle
> this change. It reduces file system stress, makes it easier for us to
> navigate the directory where we run simulations, and there is no
> performance penalty (in fact the I/O performance may improve).
>
> The only reason we wouldn't make this the default behavior is if we're
> concerned not having separate diagnostic files will impact our work-flows
> as we struggle to adapt to new pgkyl commands to fetch our preferred moment
> from within the grid or integrated diagnostics file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ammarhakim/gkyl/pull/64#issuecomment-849812508>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQ6DKHD6D55JEMVVRI4OF3TPZ62TANCNFSM45S34JKQ>
> .
>

nmandell commented on App/Diagnostics/GkDiagnostics.lua:
```diff
@@ -0,0 +1,457 @@
+-- Gkyl ------------------------------------------------------------------------
+--
+-- Diagnostics for VlasovSpecies.
```
Should be GkSpecies

nmandell reviewed (commented):
Overall this is really strong work. App-ifying the diagnostics will make adding/maintaining diagnostics much easier. It looks like it will be very straightforward for anyone to add a new diagnostic by copying and modifying one of your existing recipes.

A few more detailed comments as I scroll through the actual code changes:
1.  originally `GkEnergy` was called `GkHamilEnergy`. I think I would prefer this thing being called something more descriptive like `Energy` instead of `HE`. I moved away from HamilEnergy because I thought it would be confusing to the user what this meant. ParticleEnergy is another option, but I think simply Energy is descriptive enough. (also I wanted the boundary flux of this moment to be EnergyFlux, which seems sufficiently descriptive for users to identify this with what they want, which is the heat flux)
2.  it’s not immediately clear to me what `volIntegral.comps1` and `volIntegral.compsVdim` are. perhaps these could be named more descriptively. perhaps `volIntegral.scalar` and `volIntegral.vector`
3.  I’m confused by most of the delta F stuff. for example, in `GkSpecies:getNumDensity`, you have replaced
```
if self.deltaF then fIn:accumulate(-1.0, self.f0) end
```
with
```
self.returnFluctuation(fIn)
```
and then you compute the density with
```
self.numDensityCalc:advance(nil, {fIn}, { self.numDensityAux })
```
However, `returnFluctuation` doesn’t actually operate on `fIn`, it returns some separate field, and so `fIn` stays the same.
4.  I also don’t like how in `calcCouplingMoments`, you call a function called `returnDeltaF`, which is confusing in the most common (non-delta-f) case in which this function does the opposite of what it is named. So here, I think it would be better to simply call this `getF` or `returnF` (some places you use `get` and other places `return`, and I find this also confusing, you should probably just use one or the other)
5.  Is there a reason for having `returnFluctuation` in `GkSpecies` and then `returnDeltaF` in `KineticSpecies`, which look like they do the same thing?

manauref commented:
Addressing comments by @nmandell :

1. We'll go with `Energy` and `intEnergy`. This is the total particle energy (including potential energy) and it is clear because it is a species diagnostics and its file has the species name as a prefix.
2. I've switched `comps1`->`scalar` and `compsN`,`comsVdim`->`vector`.
3. I think there was a legitimate bug there. It might've been a benign bug in my tests but it could've caused problem elsewhere. The fix is to call `returnFluctuation` (now `getF_or_deltaF`) with an assignment: `fIn = self.getF_or_deltaF(fIn)`.
4. We have gone with `getF_or_deltaF` for clarity and to not use a name too close to the species method `getDistF`.
5. This is legacy from FluidSpecies/GyrofluidSpecies in which case different functions are necessary, but that is not the case in kinetics. I've done away with that complexity and now we simply use `getF_or_deltaF`.
